### PR TITLE
Add fungible entangler to token offerings

### PR DESCRIPTION
**Pull request by ThornM9**

### CODE REVIEW

File: packages/marketplace-ui/pages/swap/[id]/index.tsx [hunk 1]
```diff
@@ -40,7 +43,7 @@ export const SwapDisplay: NextPage = ({
         title={`Strata Swap | ${name}`}
         description={description}
         image={image}
-        url={`${SITE_URL}/bounty/${mintKey}/`}
+        url={`${SITE_URL}/bounty/${ parentEntangler ? parentEntangler.parentMint.toString() : tokenBonding?.targetMint.toString()}/`}
```
Comment (ChewingGlass) [inline packages/marketplace-ui/pages/swap/[id]/index.tsx:46]: Oh.. I think this url was already off. Should be /swap right?

File: packages/react/src/components/Swap/Swap.tsx [hunk 2]
```diff
@@ -33,7 +54,51 @@ export const Swap = ({ tokenBondingKey }: { tokenBondingKey: PublicKey }) => {
...
+          if (!isFirst || !childEntangler || !parentEntangler || isBuy) {
+            return {
+              instructions: [],
+              signers: [],
+              output: null,
+            }
+          }
+
+          let numAmount = toNumber(amount!, childMint)
+          const instr = await fungibleEntanglerSdk?.swapParentForChildInstructions({
```
Comment (ChewingGlass) [inline packages/react/src/components/Swap/Swap.tsx:68]: Could probably just add fungibleEntanglerSdk to the if statement above and simplify this to a return.

File: packages/react/src/components/Swap/Swap.tsx [hunk 3]
```diff
@@ -33,7 +54,51 @@ export const Swap = ({ tokenBondingKey }: { tokenBondingKey: PublicKey }) => {
...
+        async postInstructions({isLast, amount, isBuy}: IPostInstructionArgs): Promise<InstructionResult<null>> {
+          if (!isLast || !childEntangler || !parentEntangler || !isBuy) {
+            return {
+              instructions: [],
+              signers: [],
+              output: null,
+            }
+          }
+          const numAmount = toNumber(amount!, childMint)
+          const instr = await fungibleEntanglerSdk?.swapChildForParentInstructions({
```
Comment (ChewingGlass) [inline packages/react/src/components/Swap/Swap.tsx:88]: same here

File: packages/react/src/hooks/useSwap.ts [hunk 4]
```diff
@@ -2,16 +2,13 @@ import { useAsyncCallback } from "react-async-hook";
...
 import { BN } from "@project-serum/anchor";
 
 export type SwapArgs = {
-  extraInstructions?: (args: {
-    tokenBonding: ITokenBonding;
-    isBuy: boolean;
-    amount: BN | undefined;
-  }) => Promise<InstructionResult<null>>;
+  extraInstructions?: (args: IExtraInstructionArgs) => Promise<InstructionResult<null>>; // instructions executed before swap instructions
+  postInstructions?: (args: IPostInstructionArgs) => Promise<InstructionResult<null>>; // instructions executed after swap instructions
```
Comment (ChewingGlass) [inline packages/react/src/hooks/useSwap.ts:11]: preInstructions and postInstructions? And then roll extraInstructions (and deprecate) into pre to not be breaking?

File: packages/react/src/components/Swap/Swap.tsx [hunk 5]
```diff
@@ -33,7 +54,51 @@ export const Swap = ({ tokenBondingKey }: { tokenBondingKey: PublicKey }) => {
     tradingMints,
     onTradingMintsChange: setTradingMints,
     swap: (args) =>
-      execute(args).then(({ targetAmount }) => {
+      execute({
+        async extraInstructions({isFirst, amount, isBuy}: IExtraInstructionArgs): Promise<InstructionResult<null>> {
```
Comment (ChewingGlass) [inline packages/react/src/components/Swap/Swap.tsx:58]: If swap driver takes an id, shouldn't these extra instrs be added in useSwapDriver?

File: packages/spl-token-bonding/src/bondingHierarchy.ts [hunk 6]
```diff
@@ -72,6 +77,35 @@ export class BondingHierarchy {
...
+        hierarchy.tokenBonding.baseMint.equals(
+          sanitizeSolMint(one, this.wrappedSolMint)
+        ) ||
+        hierarchy.tokenBonding.baseMint.equals(
+          sanitizeSolMint(two, this.wrappedSolMint)
+        )
+    )?.tokenBonding?.baseMint;
+  }
+
+  highest(one: PublicKey, two: PublicKey, childEntangledMint?: PublicKey | undefined, parentEntangledMint?: PublicKey | undefined): PublicKey {
```
Comment (ChewingGlass) [inline packages/spl-token-bonding/src/bondingHierarchy.ts:92]: Is this something that would be better with composition? I.e. you create an entangled bonding hierarchy that calls out to the bonding hierarchy but makes these assumptions? Or via inheritance? Feels odd to have them as args.

**Review by ChewingGlass — CHANGES_REQUESTED** [review]

**Comment by bryzettler** [comment]
Address what @ChewingGlass brought up and make sure tests pass and I think we're good. Still need to decide when we want to roll a new release

File: packages/spl-token-bonding/src/bondingHierarchy.ts (hunk 6 above)
Comment (ThornM9) [reply packages/spl-token-bonding/src/bondingHierarchy.ts:92]: Yeah probably better for separation of concerns, will do

**Review by ChewingGlass — APPROVED** [review]

**Pull request merged**